lib/services/local/usagereporter.go

```diff
@@ -0,0 +1,544 @@
+			ClusterName: clusterName,
+			Timestamp:   time,
+			Event: &prehogapi.SubmitEventRequest_UiOnboardRecoveryCodesContinueClick{
+				UiOnboardRecoveryCodesContinueClick: (*prehogapi.UIOnboardRecoveryCodesContinueClickEvent)(e),
+			},
+		}, nil
+	default:
+		return nil, trace.BadParameter("unexpected event usage type %T", event)
+	}
+}
+
+func (r *UsageReporter) SubmitAnonymizedUsageEvents(events ...services.UsageAnonymizable) error {
+	var anonymized []*prehogapi.SubmitEventRequest
+
+	for _, e := range events {
+		e.Anonymize(r.anonymizer)
+
+		converted, err := r.convertEvent(e)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		anonymized = append(anonymized, converted)
+
+		usageEventsSubmitted.Inc()
+	}
+
+	r.events <- anonymized
+
+	return nil
+}
+
+// resubmitEvents resubmits events that have already been processed (in case of
+// some error during submission).
+func (r *UsageReporter) resubmitEvents(events []*prehogapi.SubmitEventRequest) {
+	usageEventsRequeuedTotal.Add(float64(len(events)))
+
+	r.events <- events
+}
+
+func NewPrehogSubmitter(ctx context.Context, prehogEndpoint string, clientCert *tls.Certificate, caCertPEM []byte) (UsageSubmitFunc, error) {
+	tlsConfig := &tls.Config{
+		// Self-signed test licenses may not have a proper issuer and won't be
+		// used if just passed in via Certificates, so we'll use this to
+		// explicitly set the client cert we want to use.
+		GetClientCertificate: func(cri *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return clientCert, nil
+		},
+	}
+
+	if caCertPEM != nil {
+		pool := x509.NewCertPool()
+		pool.AppendCertsFromPEM(caCertPEM)
+
+		tlsConfig.RootCAs = pool
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			ForceAttemptHTTP2:   true,
+			Proxy:               http.ProxyFromEnvironment,
+			TLSClientConfig:     tlsConfig,
+			IdleConnTimeout:     defaults.HTTPIdleTimeout,
+			MaxIdleConns:        defaults.HTTPMaxIdleConns,
+			MaxIdleConnsPerHost: defaults.HTTPMaxConnsPerHost,
+		},
+		CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+		Timeout: 5 * time.Second,
+	}
+
+	client := prehogclient.NewTeleportReportingServiceClient(httpClient, prehogEndpoint)
+
+	return func(reporter *UsageReporter, events []*prehogapi.SubmitEventRequest) error {
+		// Note: the backend doesn't support batching at the moment.
+		for _, event := range events {
+			// Note: this results in retrying the entire batch, which probably
+			// isn't ideal.
+			req := connect.NewRequest(event)
+			if _, err := client.SubmitEvent(ctx, req); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		return nil
+	}, nil
+}
+
+// NewUsageReporter creates a new usage reporter. `Run()` must be executed to
+// process incoming events.
+func NewUsageReporter(ctx context.Context, parentLog logrus.FieldLogger, clusterName types.ClusterName, submitter UsageSubmitFunc) (*UsageReporter, error) {
```

Comment:
Similar to my previous comment. `NewUsageReporter` is not a long running operation that needs to support cancelation. `Run` is the long running operation.

I would move the context from the constructor into `Run`.

Reply:
that feels much cleaner, done

(`cancel()` wasn't even used... oops)

Reply:
Looks much better, thanks!